Fix HPA sample sanitization

jbartosik opened the pull request:
**What this PR does / why we need it**: @mwielgus pointed out a case when HPA fails as a result of my changes to HPA algorithm:
- Have pods that use a lot of CPU during initilization, become ready right after they initialize,
- Trigger a scale up,
- When new pods become ready will will count their usage (even though it's not related to any work that needs doing),
- This triggers another scale up, even though existing pods can handle work, no problem.

The fix is:
- Use all samples for non-cpu metrics.
- Only use CPU samples if:
  - Pod is ready and was started more than 2 minutes ago, or
  - Pod is unready and last readiness change happened more than 10s after it was started.

Reasoning behind this in: https://docs.google.com/document/d/1UdtYedhmCxjaJIQi6hwJMY0eHQQKxlVD8lSHZC1BPOA/edit

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Release note**:
```release-note
Replace scale up forbidden window with disregarding CPU samples collected when pod was initializing.
```

jbartosik committed:
Improve HPA sample sanitization

After my previous changes HPA wasn't behaving correctly in the following
situation:

- Pods use a lot of CPU during initilization, become ready right after they initialize,
- Scale up triggers,
- When new pods become ready HPA counts their usage (even though it's not related to any work that needs doing),
- Another scale up, even though existing pods can handle work, no problem.

MaciekPytel commented:
/ok-to-test

jbartosik commented:
/retest

jbartosik commented:
/retest

jbartosik commented:
/retest

jbartosik commented:
/retest

DirectXMan12 reviewed (changes requested):
you need to add what the fix is to the PR description.  This definitely needs a release note.  All algorithm changes need release notes.

Separately, hard coding a startup window is not acceptable.  Some pods start up within 10s, others take 5 minutes.  We can't just arbitrarily assume 2 minutes is correct.

jbartosik commented:
Done. I Updated description. And also I changed implementation a little (to match the conclusions in the doc).

jbartosik commented:
@DirectXMan12 please take a look

jbartosik committed:
Make HPA more configurable

Duration of initialization taint on CPU and window of initial readiness
setting controlled by flags.

Adding API violation exceptions following example of e50340ee238

jbartosik commented:
@DirectXMan12 I added a commit turning hard coded durations into flags.

I rebased because the new commit had a merge conflict.

jbartosik commented:
/assign @lavalamp

DirectXMan12 commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -29,29 +30,32 @@ import (
-	// less than maxDelayOfInitialReadinessStatus after pod start we assume it has never been ready.
-	maxDelayOfInitialReadinessStatus = 10 * time.Second
+	defaultTestingTolerance                     = 0.1
+	defaultTestingCpuTaintAfterStart            = 2 * time.Minute
```
a) you need a comment as to why this is two minutes
b) if this is based on the metrics-server collection window, it shouldn't be constant -- it should be using the reported window size from metrics-server.  If this is "just a guess", that's not a good answer.

jbartosik replied:
thought you left this comment in a different commit and were referring to
values HPA uses (which were hard coded in the previous commit).

Those are just values I use for testing, it doesn't matter much what exactly they are.

jbartosik commented:
/retest

jbartosik commented:
/retest

jbartosik commented:
/retest

DirectXMan12 commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -416,6 +416,14 @@ type HPAControllerConfiguration struct {
+	// changes are treated as readiness being set for the first time. The only effect of this is that
+	// HPA will disregard CPU samples from unready pods that had last readiness change during that
+	// period.
+	HorizontalPodAutoscalerInitialReadinessDelay metav1.Duration
```
is this actually based on the metric window?  if so, it should use the metric window reported by the custom metrics API, not be hardcoded.

jbartosik replied:
I agree in principle. But that would be significantly more complicated (currently client doesn't return metric window so I would have to change client, add more tests, and modify existing tests).

I need to merge this in 2 days so I'd rather not do it in this PR. Can we settle for a TODO. I'd send a PR addressing the TODO when I'm back from vacation (and merge it after the freeze).

DirectXMan12 commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -386,21 +342,60 @@ func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(currentReplicas int3
+			if condition == nil || pod.Status.StartTime == nil {
+				ignorePod = true
+			} else {
+				glog.Errorf("cpuTaintAfterStart: %v; delayOfInitialReadinessStatus: %v", cpuTaintAfterStart, delayOfInitialReadinessStatus)
```
extraneous log statement.

DirectXMan12 reviewed (changes requested):
found an extraneous log line.  modulo that and the changes around window calculation we discussed, this looks good.

jbartosik commented:
/retest

mwielgus commented on cmd/kube-controller-manager/app/options/options_test.go:
```diff
@@ -73,6 +73,8 @@ func TestAddFlags(t *testing.T) {
 		"--horizontal-pod-autoscaler-sync-period=45s",
 		"--horizontal-pod-autoscaler-upscale-delay=1m",
+		"--horizontal-pod-autoscaler-cpu-taint-period=90s",
+		"--horizontal-pod-autoscaler-initial-readiness-delay=5s",
```
5s?

mwielgus commented on cmd/kube-controller-manager/app/options/options_test.go:
```diff
@@ -185,6 +187,8 @@ func TestAddFlags(t *testing.T) {
 			HorizontalPodAutoscalerUpscaleForbiddenWindow:   metav1.Duration{Duration: 1 * time.Minute},
 			HorizontalPodAutoscalerDownscaleForbiddenWindow: metav1.Duration{Duration: 2 * time.Minute},
+			HorizontalPodAutoscalerCPUTaintPeriod:           metav1.Duration{Duration: 90 * time.Second},
+			HorizontalPodAutoscalerInitialReadinessDelay:    metav1.Duration{Duration: 5 * time.Second},
```
5s?

mwielgus commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -95,6 +95,14 @@ func SetDefaults_KubeControllerManagerConfiguration(obj *KubeControllerManagerCo
+		obj.HPAController.HorizontalPodAutoscalerCPUTaintPeriod = metav1.Duration{Duration: time.Minute}
+	}
+	if obj.HPAController.HorizontalPodAutoscalerInitialReadinessDelay == zero {
+		obj.HPAController.HorizontalPodAutoscalerInitialReadinessDelay = metav1.Duration{Duration: 10 * time.Second}
```
10 sec only?

jbartosik replied:
Yes. This is what we assume is max duration between pod starting and the first time it get readiness assigned

jbartosik replied:
It was <2s for the pods I looked at

mwielgus replied:
Let's make it 30second, just in case Kubelet is stalled.

DirectXMan12 commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -98,7 +98,7 @@ func SetDefaults_KubeControllerManagerConfiguration(obj *KubeControllerManagerCo
 		// Assuming CPU is collected every minute and initialization takes another minute HPA should
 		// disregard samples from first two minutes as contaminated by initialization.
-		obj.HPAController.HorizontalPodAutoscalerCPUTaintPeriod = metav1.Duration{Duration: 2 * time.Minute}
+		obj.HPAController.HorizontalPodAutoscalerCPUTaintPeriod = metav1.Duration{Duration: time.Minute}
```
if this is based on HPA processing interval + window, and we are going to actually collect window from the API, why do we need this separate option?  Why not just use the HPA processing window, so we don't have this option at all?  (that's what I was trying to suggest before)

jbartosik replied:
We want to avoid using samples with window overlapping with pod initialization. If we use sample after {end of initialization time} + {sample window} we will achieve that.

{sample window} is hardcoded for now, I'll send PR for using actual value tomorrow.

We cannot know {end of initialization time} without storing per-pod information which is not feasible. So I'm going with a heuristic - initialization takes no more than a HorizontalPodAutoscalerCPUTaintPeriod (1 minute by default).

mwielgus commented:
/lgtm

mwielgus commented:
/approve

jbartosik commented:
/assign @wojtek-t

mwielgus commented:
@DirectXMan12 
I took a liberty of LGTMing the PR to get it through the door, based on @jbartosik claims of your verbal conditional LGTM. As @jbartosik is going for vacation next week @krzysztof-jastrzebski will be doing all of the follow-up PRs to adjust readiness-related logic to both your and our preferences.

jbartosik commented:
@wojtek-t Done.

wojtek-t commented:
/lgtm

[Though those validation exceptions need api-reviewer approval].

wojtek-t commented:
/hold

@thockin - for final approval of the exceptions.

thockin commented:
I am not sure what I am asked to approve?  what is triggering the violation?

jbartosik commented:
@thockin  I'm adding new flags for HPA. They require corresponding fields in componentconfig v1alpha1 API. All such fields have an exception (I think because they start with a capital letter).

thockin commented:
/approve

jbartosik commented:
@thockin - please remove hold on this PR @wojtek-t placed it to make sure you approve it.

jbartosik commented:
/hold cancel

jbartosik commented:
@DirectXMan12 @mwielgus please merge this PR

DirectXMan12 commented:
> As @jbartosik is going for vacation next week @krzysztof-jastrzebski will be doing all of the follow-up PRs to adjust readiness-related logic to both your and our preferences.

As long as the PRs are posted before code slush that's acceptable.
